rest/server.go

```diff
@@ -0,0 +1,108 @@
+package rest
+
+import (
+	"awesomeProject3/service"
+	"context"
+	"encoding/json"
+	"github.com/go-chi/chi/v5"
+	"log"
+	"net/http"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+)
+
+type Handler func(w http.ResponseWriter, r *http.Request) error
+
+func (h Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if err := h(w, r); err != nil {
+		w.WriteHeader(503)
+	}
+}
+
+type container struct {
+	Getter string `json:"getter"`
+	Value  string `json:"value"`
+}
+
+type SecretRestAPI struct {
+	ssService  service.SecretService
+	httpServer *http.Server
+}
+
+func NewSecretRestAPI(sr service.SecretService, portnumber string) *SecretRestAPI {
+	a := SecretRestAPI{}
+
+	h := chi.NewRouter()
+	h.Method("GET", "/", Handler(a.handlerGet))
+	h.Method("POST", "/", Handler(a.handlerPost))
+
+	a.ssService = sr
+	a.httpServer = &http.Server{Addr: "localhost:" + portnumber, Handler: h}
+
+	return &a
+}
+
+func (sr *SecretRestAPI) Start() {
+	serverCtx, serverStopCtx := context.WithCancel(context.Background())
+
+	sign := make(chan os.Signal, 1)
+	signal.Notify(sign, syscall.SIGHUP, syscall.SIGINT, syscall.SIGTERM, syscall.SIGQUIT)
+	go func() {
+		<-sign
+		log.Println("got termination signal. Try to shutdown gracefully")
+		shutdownCtx, _ := context.WithTimeout(serverCtx, 30*time.Second)
+
+		err := sr.httpServer.Shutdown(shutdownCtx)
+		if err != nil {
+			log.Fatal(err)
+		}
+		serverStopCtx()
+
+	}()
+
+	err := sr.httpServer.ListenAndServe()
+	if err != nil && err != http.ErrServerClosed {
+		log.Fatal(err)
+	}
+	log.Println("server stopped successfully")
+	<-serverCtx.Done()
+	log.Println("Done")
+}
+
+func (sr *SecretRestAPI) handlerGet(w http.ResponseWriter, r *http.Request) error {
+	key := r.URL.Query().Get("getter")
+
+	result, err := sr.ssService.ReadSecret(key, r.Header.Get("X-Cipher"))
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusNotFound)
+		return err
+	}
+	_, err = w.Write([]byte(result))
+	if nil != err {
+		http.Error(w, err.Error(), http.StatusBadRequest)
```

Comment:
What does this method do? How does this in sync with 

```go
func (h Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	if err := h(w, r); err != nil {
 		w.WriteHeader(503)
 	}
 }
```

Reply:
1) this method fills http response with simpleSecretService result data.  
2) ServeHTTP is called by router, while http request processing. then inside ServeHTTP method "h" is called which is function we passed while mapping handler methods.

> What does this method do? How does this in sync with
> 
> ```go
> func (h Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
>  	if err := h(w, r); err != nil {
>  		w.WriteHeader(503)
>  	}
>  }
> ```

Reply:
I mean that it would write `503` status code for all handler responses that return error. It doesn't sound right from my perspective.

Reply:
1) changed 503 to 500 just to be more accurate 
2) do not return error from places where ive already processed the error 
3) did not change signature of method because that is defined by external library